File: modules/image_preprocess/ip_test.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <string>
#include <iostream>
#include <fstream>

// config
#define IMAGE_SIZE 384
#define NUM_CHANNELS 3
#define rescale_factor 0.00392156862745098

// Note: change to double
// float => double & CV_32FC3 => CV_64FC3 & Vec3f => Vec3d

void normalize(float out[NUM_CHANNELS][IMAGE_SIZE][IMAGE_SIZE], std::vector<cv::Mat> channels) {
    // 假設 image 是 CV_32FC3 格式，且值域為 [0, 1]
    const float mean[NUM_CHANNELS] = {0.5, 0.5, 0.5};
    const float std[NUM_CHANNELS]  = {0.5, 0.5, 0.5};

    for (int c = 0; c < NUM_CHANNELS; ++c) {
        for (int i = 0; i < IMAGE_SIZE; ++i) {
            for (int j = 0; j < IMAGE_SIZE; ++j) {
                float pixel = channels[c].at<float>(i, j);
                out[c][i][j] = (pixel - mean[c]) / std[c];
            }
        }
    }
}

void preprocess(float out[NUM_CHANNELS][IMAGE_SIZE][IMAGE_SIZE], const cv::Mat& image_bgr) {
    // 1. 將 BGR 轉為 RGB
    cv::Mat image_rgb;
    cv::cvtColor(image_bgr, image_rgb, cv::COLOR_BGR2RGB);
    std::cout << "Before preprocess shape: (" << image_rgb.cols << ", " << image_rgb.rows << ", " << image_rgb.channels();
    std::cout << "), first pixel value: " << image_rgb.at<cv::Vec3b>(0, 0) << std::endl;

    // 2. 調整大小為 384x384
    cv::Mat resized;
    cv::resize(image_rgb, resized, cv::Size(IMAGE_SIZE, IMAGE_SIZE), 0 , 0, cv::INTER_AREA);
    std::cout << "Resized shape: (" << resized.cols << ", " << resized.rows << ", " << resized.channels();
    std::cout << "), first pixel value: " << resized.at<cv::Vec3b>(0, 0) << std::endl;

    // 3. 調整至 [0, 1] 範圍
    cv::Mat rescaled;
    resized.convertTo(rescaled, CV_32FC3, 1.0 / 255.0);
    std::cout << "Rescaled shape: (" << rescaled.cols << ", " << rescaled.rows << ", " << rescaled.channels();
    std::cout << "), first pixel value: [" << rescaled.at<cv::Vec3f>(0, 0)[0] << ", " << rescaled.at<cv::Vec3f>(0, 0)[1] << ", " << rescaled.at<cv::Vec3f>(0, 0)[2] << "]" << std::endl;

    // 4. 拆分通道（R, G, B)
    std::vector<cv::Mat> channels(3);
    cv::split(rescaled, channels);
    std::cout << "split shape: (" << 3 << ", " << channels[0].cols << ", " << channels[0].rows;
    std::cout << "), first pixel of each channel: " << channels[0].at<float>(0, 0) << ", " << channels[1].at<float>(0, 0) << ", " << channels[2].at<float>(0, 0) << std::endl;

    // 5. 正規化   
    normalize(out, channels);
    std::cout << "normalize shape: (" << NUM_CHANNELS << ", " << IMAGE_SIZE << ", " << IMAGE_SIZE;
    std::cout << "), first pixel of each channel: " << out[0][0][0] << ", " << out[1][0][0] << ", " << out[2][0][0] << std::endl;
}

int main() {
    std::string image_path = "test.jpg";

    // load image in BGR & HWC format => [Height(=image.rows), Width(=image.cols), Channels]
    cv::Mat img = cv::imread(image_path, cv::IMREAD_COLOR);
    if (img.empty()) {
        std::cerr << "Failed to load image from file: " << image_path << "\n";
        return 1;
    }

    // preprocess the image
    float preprocessed_img[NUM_CHANNELS][IMAGE_SIZE][IMAGE_SIZE];
    preprocess(preprocessed_img, img);
    std::cout << "After preprocess shape: (3, 384, 384)";
    std::cout << ", first pixel value: " << preprocessed_img[0][0][0] << ", " << preprocessed_img[1][0][0] << ", " << preprocessed_img[2][0][0] << std::endl;
    
    // Save the preprocessed image to a binary file
    std::string filename = "preprocessed_implementation_cpp.bin";
    std::ofstream ofs(filename, std::ios::binary);
    if (!ofs) {
        std::cerr << "Cannot open file for writing: " << filename << std::endl;
        return 0;
    }
    ofs.write(reinterpret_cast<const char*>(preprocessed_img), sizeof(float) * NUM_CHANNELS * IMAGE_SIZE * IMAGE_SIZE);
    ofs.close();

    return 0;
}

File: modules/image_preprocess/ip_test.py
from transformers import SiglipImageProcessor
from transformers.image_processing_utils import rescale, normalize
import requests
from PIL import Image
from io import BytesIO
import numpy as np
import cv2

# config
image_mean = [0.5, 0.5, 0.5]
image_std = [0.5, 0.5, 0.5]
resample = 3
rescale_factor = 0.00392156862745098
(height, width) = (384, 384)

# Option1: load image from URL
# image_url = "http://images.cocodataset.org/test-stuff2017/000000000001.jpg"
# response = requests.get(image_url)
# image = Image.open(BytesIO(response.content)).convert("RGB")
# print(f"image size: {image.size}, first pixel value: {image.getpixel((0, 0))}")

# Option2: load image from local file
# image = Image.open("test.jpg").convert("RGB")
# print(f"before shape: ({image.width}, {image.height}, 3), first pixel value: {image.getpixel((0, 0))}")

# SiglipImageProcessor
print("***SiglipImageProcessor***")
cv_img = cv2.imread("test.jpg", cv2.IMREAD_COLOR)
cv_img_rgb = cv2.cvtColor(cv_img, cv2.COLOR_BGR2RGB)
image = Image.fromarray(cv_img_rgb)
print(f"Before preprocess shape: ({image.height}, {image.width}, 3), first pixel value: {image.getpixel((0, 0))}")
processor = SiglipImageProcessor.from_pretrained("google/siglip-so400m-patch14-384")
image = processor(images=[image], return_tensors='pt')['pixel_values']
print(f"After preprocess shape: {image.shape}, first pixel value: {image[0][0][0][0]}, {image[0][1][0][0]}, {image[0][2][0][0]}\n")
image.squeeze(0).numpy().tofile("preprocessed_SiglipImageProcessor.bin")

# Implementation of SiglipImageProcessor (without batch dimension)
print("***Implementation of SiglipImageProcessor (without batch dimension)***")
cv_img = cv2.imread("test.jpg", cv2.IMREAD_COLOR)
cv_img_rgb = cv2.cvtColor(cv_img, cv2.COLOR_BGR2RGB)
image = Image.fromarray(cv_img_rgb)
print(f"Before preprocess shape: ({image.height}, {image.width}, 3), first pixel value: {image.getpixel((0, 0))}")
image = image.resize(size=(width, height), resample=resample)
image = np.array(image)
print(f"Resized shape: {image.shape}, first pixel value: {image[0][0]}")
image = rescale(image, scale=rescale_factor)
print(f"Rescaled shape: {image.shape}, first pixel value: {image[0][0]}")
image = normalize(image, mean=image_mean, std=image_std)
print(f"Normalized shape: {image.shape}, first pixel value: {image[0][0]}")
axes = list(range(image.ndim - 3)) + [image.ndim - 1, image.ndim - 3, image.ndim - 2]
image = image.transpose(axes)
print(f"After preprocess shape: {image.shape}, first pixel value: {image[0][0][0]}, {image[1][0][0]}, {image[2][0][0]}")
image.tofile("preprocessed_implementation_py.bin")

# Compare with SiglipImageProcessor
print("\n===Comparison Summary===")
bin1 = np.fromfile("preprocessed_SiglipImageProcessor.bin", dtype=np.float32)
# bin2 = np.fromfile("preprocessed_implementation_py.bin", dtype=np.float32)
bin2 = np.fromfile("preprocessed_implementation_cpp.bin", dtype=np.float32)
if bin1.shape != bin2.shape:
    print(f"Shape mismatch: bin1 {bin1.shape}, bin2 {bin2.shape}")
else:
    print(f"Compare {bin1.shape} values\n")
    print("First 10 values of bin1:\n", bin1[:10])
    print("First 10 values of bin2:\n", bin2[:10])

    abs_err = np.abs(bin1 - bin2)
    rel_err = abs_err / (np.abs(bin1) + 1e-12)

    print("\nAbsolute Error:")
    print(" - Maximum:", np.max(abs_err))
    print(" - Average:", np.mean(abs_err))

    print("\nRelative Error:")
    print(f" - Maximum: {np.max(rel_err)} %")
    print(f" - Average: {np.mean(rel_err)} %")

    # abs_threshold = 0.1
    # rel_threshold = 0.05
    # abs_large_idx = np.where(abs_err > abs_threshold)[0]
    # rel_large_idx = np.where(rel_err > rel_threshold)[0]
    # print(f"\nIndices with absolute error > {abs_threshold} (total {len(abs_large_idx)}):")
    # for idx in abs_large_idx:
    #     print(f"Index {idx}: bin1={bin1[idx]}, bin2={bin2[idx]}, abs_err={abs_err[idx]:.6f}")
    # print(f"\nIndices with relative error > {rel_threshold} (total {len(rel_large_idx)}):")
    # for idx in rel_large_idx:
    #     print(f"Index {idx}: bin1={bin1[idx]}, bin2={bin2[idx]}, rel_err={rel_err[idx]:.6f}")

